File: src/memory/models/skill_learning.py
"""Durable Skill learning control-plane models.

These tables do not create a second Memory authority. They persist proposal,
usage-evidence, and append-only proposal-history records for the existing Skill
system. Live Skill mutation remains file-backed and explicit-apply only.
"""

from __future__ import annotations

import json
import uuid
from datetime import datetime
from typing import Any, Dict

from sqlalchemy import (
    CheckConstraint,
    Column,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    JSON,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from .base import Base, _encrypted_json_property, _encrypted_text_property
from ...security.skill_content_privacy import redact_secret_like_skill_content


def _public_provenance(value: Any) -> Any:
    """Return provenance without leaking legacy/plaintext secret material."""
    candidate = value
    if isinstance(candidate, str):
        stripped = candidate.strip()
        if not stripped:
            return {}
        try:
            candidate = json.loads(stripped)
        except (TypeError, ValueError, json.JSONDecodeError):
            candidate = stripped
    if candidate in (None, ""):
        return {}
    return redact_secret_like_skill_content(candidate)


class SkillUsageReceipt(Base):
    """Evidence that one concrete Skill was actually invoked."""

    __tablename__ = "skill_usage_receipts"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(String(200), nullable=False, index=True)
    project_id = Column(
        UUID(as_uuid=True),
        ForeignKey("projects.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    session_id = Column(
        UUID(as_uuid=True),
        ForeignKey("conversation_sessions.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    message_id = Column(
        UUID(as_uuid=True),
        ForeignKey("conversation_messages.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    agent_run_id = Column(
        UUID(as_uuid=True),
        ForeignKey("agent_runs.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    tool_call_id = Column(String(160), nullable=True, index=True)
    invocation_path = Column(String(32), nullable=False, index=True)
    outcome = Column(String(16), nullable=False, index=True)
    skill_name = Column(String(160), nullable=False, index=True)
    skill_scope = Column(String(16), nullable=False, index=True)
    skill_path = Column(String(512), nullable=False)
    skill_hash = Column(String(64), nullable=False, index=True)
    skill_version = Column(String(80), nullable=False)
    idempotency_key = Column(String(128), nullable=False, unique=True, index=True)
    _provenance = Column("provenance", Text, nullable=False, default="")
    provenance = _encrypted_json_property(
        "_provenance",
        "skill_usage_receipts.provenance",
    )
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow, index=True)

    __table_args__ = (
        CheckConstraint(
            "skill_scope IN ('global', 'project')",
            name="ck_skill_usage_receipts_scope",
        ),
        CheckConstraint(
            "outcome IN ('success', 'error')",
            name="ck_skill_usage_receipts_outcome",
        ),
        Index(
            "ix_skill_usage_receipts_skill_created",
            "skill_scope",
            "skill_name",
            "created_at",
        ),
    )

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": str(self.id),
            "user_id": self.user_id,
            "project_id": str(self.project_id) if self.project_id else None,
            "session_id": str(self.session_id) if self.session_id else None,
            "message_id": str(self.message_id) if self.message_id else None,
            "agent_run_id": str(self.agent_run_id) if self.agent_run_id else None,
            "tool_call_id": self.tool_call_id,
            "invocation_path": self.invocation_path,
            "outcome": self.outcome,
            "skill_name": self.skill_name,
            "skill_scope": self.skill_scope,
            "skill_path": self.skill_path,
            "skill_hash": self.skill_hash,
            "skill_version": self.skill_version,
            "provenance": _public_provenance(self.provenance),
            "created_at": self.created_at.isoformat() if self.created_at else None,
        }


class SkillProposal(Base):
    """Proposal-first mutation record for one canonical Skill target."""

    __tablename__ = "skill_proposals"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(String(200), nullable=False, index=True)
    project_id = Column(
        UUID(as_uuid=True),
        ForeignKey("projects.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    receipt_id = Column(
        UUID(as_uuid=True),
        ForeignKey("skill_usage_receipts.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    operation = Column(String(16), nullable=False, index=True)
    reason_type = Column(String(24), nullable=False, default="manual", index=True)
    target_scope = Column(String(16), nullable=False, index=True)
    target_name = Column(String(160), nullable=False, index=True)
    target_path = Column(String(512), nullable=False)
    status = Column(String(16), nullable=False, default="pending", index=True)
    proposed_content = Column(JSON, nullable=False, default=dict)
    base_snapshot = Column(JSON, nullable=True)
    _base_text = Column("base_text", Text, nullable=True)
    base_text = _encrypted_text_property(
        "_base_text",
        "skill_proposals.base_text",
    )
    base_hash = Column(String(64), nullable=True, index=True)
    base_version = Column(String(80), nullable=True)
    applied_hash = Column(String(64), nullable=True, index=True)
    applied_version = Column(String(80), nullable=True)
    idempotency_key = Column(String(128), nullable=False, unique=True, index=True)
    _provenance = Column("provenance", Text, nullable=False, default="")
    provenance = _encrypted_json_property(
        "_provenance",
        "skill_proposals.provenance",
    )
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow, index=True)
    updated_at = Column(
        DateTime,
        nullable=False,
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
    )
    applied_at = Column(DateTime, nullable=True)
    rejected_at = Column(DateTime, nullable=True)
    stale_at = Column(DateTime, nullable=True)
    rolled_back_at = Column(DateTime, nullable=True)

    receipt = relationship("SkillUsageReceipt")
    history = relationship(
        "SkillProposalHistory",
        back_populates="proposal",
        cascade="all, delete-orphan",
        order_by="SkillProposalHistory.sequence",
        lazy="selectin",
    )

    __table_args__ = (
        CheckConstraint(
            "operation IN ('create', 'update')",
            name="ck_skill_proposals_operation",
        ),
        CheckConstraint(
            "reason_type IN ('manual', 'correction', 'procedure')",
            name="ck_skill_proposals_reason_type",
        ),
        CheckConstraint(
            "target_scope IN ('global', 'project')",
            name="ck_skill_proposals_scope",
        ),
        CheckConstraint(
            "status IN ('pending', 'applied', 'rejected', 'stale')",
            name="ck_skill_proposals_status",
        ),
        Index(
            "ix_skill_proposals_target_status",
            "target_scope",
            "project_id",
            "target_name",
            "status",
        ),
    )

    def to_dict(self, *, include_history: bool = False) -> Dict[str, Any]:
        payload: Dict[str, Any] = {
            "id": str(self.id),
            "user_id": self.user_id,
            "project_id": str(self.project_id) if self.project_id else None,
            "receipt_id": str(self.receipt_id) if self.receipt_id else None,
            "operation": self.operation,
            "reason_type": self.reason_type,
            "target_scope": self.target_scope,
            "target_name": self.target_name,
            "target_path": self.target_path,
            "status": self.status,
            "proposed_content": redact_secret_like_skill_content(
                self.proposed_content or {}
            ),
            "base_snapshot": redact_secret_like_skill_content(
                self.base_snapshot
            ),
            "base_hash": self.base_hash,
            "base_version": self.base_version,
            "applied_hash": self.applied_hash,
            "applied_version": self.applied_version,
            "provenance": _public_provenance(self.provenance),
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
            "applied_at": self.applied_at.isoformat() if self.applied_at else None,
            "rejected_at": self.rejected_at.isoformat() if self.rejected_at else None,
            "stale_at": self.stale_at.isoformat() if self.stale_at else None,
            "rolled_back_at": (
                self.rolled_back_at.isoformat() if self.rolled_back_at else None
            ),
        }
        if include_history:
            payload["history"] = [item.to_dict() for item in self.history or []]
        return payload


class SkillProposalHistory(Base):
    """Append-only audit history for a Skill proposal."""

    __tablename__ = "skill_proposal_history"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    proposal_id = Column(
        UUID(as_uuid=True),
        ForeignKey("skill_proposals.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    sequence = Column(Integer, nullable=False)
    event = Column(String(24), nullable=False, index=True)
    actor_user_id = Column(String(200), nullable=False)
    from_status = Column(String(16), nullable=True)
    to_status = Column(String(16), nullable=True)
    observed_hash = Column(String(64), nullable=True)
    result_hash = Column(String(64), nullable=True)
    _details = Column("details", Text, nullable=False, default="")
    details = _encrypted_json_property(
        "_details",
        "skill_proposal_history.details",
    )
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow, index=True)

    proposal = relationship("SkillProposal", back_populates="history")

    __table_args__ = (
        UniqueConstraint(
            "proposal_id",
            "sequence",
            name="uq_skill_proposal_history_sequence",
        ),
    )

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": str(self.id),
            "proposal_id": str(self.proposal_id),
            "sequence": self.sequence,
            "event": self.event,
            "actor_user_id": self.actor_user_id,
            "from_status": self.from_status,
            "to_status": self.to_status,
            "observed_hash": self.observed_hash,
            "result_hash": self.result_hash,
            "details": redact_secret_like_skill_content(
                self.details or {}
            ),
            "created_at": self.created_at.isoformat() if self.created_at else None,
        }

File: frontend/src/lib/skill-learning.ts
"use client";

export type SkillProposalStatus =
  | "pending"
  | "applied"
  | "rejected"
  | "stale";

export type SkillProposalOperation = "create" | "update";
export type SkillProposalScope = "global" | "project";
export type SkillProposalReason = "manual" | "correction" | "procedure";

export interface SkillProposalContent {
  name: string;
  description: string;
  prompt_template: string;
  trigger_mode: string;
  aliases: string[];
  bound_tools: string[];
  examples: string[];
  tags: string[];
  parameters: Record<string, unknown>;
}

export interface SkillProposalHistoryEntry {
  id: string;
  proposal_id: string;
  sequence: number;
  event: string;
  actor_user_id: string;
  from_status?: string | null;
  to_status?: string | null;
  observed_hash?: string | null;
  result_hash?: string | null;
  details: Record<string, unknown>;
  created_at?: string | null;
}

export interface SkillProposal {
  id: string;
  user_id: string;
  project_id?: string | null;
  receipt_id?: string | null;
  operation: SkillProposalOperation;
  reason_type: SkillProposalReason;
  target_scope: SkillProposalScope;
  target_name: string;
  target_path: string;
  status: SkillProposalStatus;
  proposed_content: SkillProposalContent;
  base_snapshot?: SkillProposalContent | null;
  base_hash?: string | null;
  base_version?: string | null;
  applied_hash?: string | null;
  applied_version?: string | null;
  provenance: Record<string, unknown>;
  created_at?: string | null;
  updated_at?: string | null;
  applied_at?: string | null;
  rejected_at?: string | null;
  stale_at?: string | null;
  rolled_back_at?: string | null;
  history?: SkillProposalHistoryEntry[];
}

export interface SkillUsageReceipt {
  id: string;
  user_id: string;
  project_id?: string | null;
  session_id?: string | null;
  message_id?: string | null;
  agent_run_id?: string | null;
  tool_call_id?: string | null;
  invocation_path: string;
  outcome: "success" | "error" | string;
  skill_name: string;
  skill_scope: SkillProposalScope;
  skill_path: string;
  skill_hash: string;
  skill_version: string;
  provenance: Record<string, unknown>;
  created_at?: string | null;
}

export interface CreateSkillProposalInput {
  operation: SkillProposalOperation;
  target_scope: SkillProposalScope;
  target_name: string;
  project_id?: string;
  reason_type?: SkillProposalReason;
  proposed_content: Partial<SkillProposalContent>;
  receipt_id?: string;
  evidence?: Array<Record<string, unknown>>;
  provenance?: Record<string, unknown>;
  idempotency_key?: string;
}

export interface SkillProposalRevisionToken {
  updated_at?: string | null;
  status: SkillProposalStatus;
  base_hash?: string | null;
  base_version?: string | null;
  applied_hash?: string | null;
  applied_version?: string | null;
  rolled_back_at?: string | null;
}

export type SkillLearningApiErrorCode =
  | "network"
  | "unauthorized"
  | "forbidden"
  | "not_found"
  | "validation"
  | "conflict"
  | "stale"
  | "server"
  | "unknown";

export class SkillLearningApiError extends Error {
  readonly status: number;
  readonly code: SkillLearningApiErrorCode;
  readonly proposal?: SkillProposal;

  constructor(
    message: string,
    status: number,
    code: SkillLearningApiErrorCode,
    proposal?: SkillProposal,
  ) {
    super(message);
    this.name = "SkillLearningApiError";
    this.status = status;
    this.code = code;
    this.proposal = proposal;
  }
}

const API_BASE = "/api/python-proxy/skills";

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function proposalFromUnknown(value: unknown): SkillProposal | undefined {
  if (
    !isRecord(value) ||
    typeof value.id !== "string" ||
    typeof value.target_name !== "string" ||
    typeof value.status !== "string"
  ) {
    return undefined;
  }
  return value as unknown as SkillProposal;
}

function errorCodeForStatus(
  status: number,
  proposal?: SkillProposal,
): SkillLearningApiErrorCode {
  if (status === 401) return "unauthorized";
  if (status === 403) return "forbidden";
  if (status === 404) return "not_found";
  if (status === 400 || status === 422) return "validation";
  if (status === 409 && proposal?.status === "stale") return "stale";
  if (status === 409) return "conflict";
  if (status >= 500) return "server";
  return "unknown";
}

function fallbackErrorMessage(status: number): string {
  if (status === 401) return "認証が必要です。";
  if (status === 403) return "この Skill 提案を操作する権限がありません。";
  if (status === 404) return "Skill 提案または使用実績が見つかりません。";
  if (status === 400 || status === 422) {
    return "Skill 提案の入力内容を確認してください。";
  }
  if (status === 409) {
    return "Skill または提案が更新されています。最新状態を確認してください。";
  }
  if (status >= 500) {
    return "Skill 提案の処理に失敗しました。再読み込みして再試行してください。";
  }
  return `Skill 提案 API エラー (${status})`;
}

async function apiError(response: Response): Promise<SkillLearningApiError> {
  const payload = await response.json().catch(() => null);
  const detail = isRecord(payload) ? payload.detail : undefined;

  let detailMessage: string | undefined;
  let proposal: SkillProposal | undefined;

  if (typeof detail === "string") {
    detailMessage = detail;
  } else if (isRecord(detail)) {
    if (typeof detail.message === "string") {
      detailMessage = detail.message;
    }
    proposal = proposalFromUnknown(detail.proposal);
  }

  const code = errorCodeForStatus(response.status, proposal);
  const allowBackendMessage =
    response.status === 400 ||
    response.status === 404 ||
    response.status === 409 ||
    response.status === 422;
  const message =
    allowBackendMessage && detailMessage
      ? detailMessage
      : fallbackErrorMessage(response.status);

  return new SkillLearningApiError(message, response.status, code, proposal);
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const { headers, ...rest } = init ?? {};
  let response: Response;
  try {
    response = await fetch(`${API_BASE}${path}`, {
      credentials: "include",
      ...rest,
      headers: {
        "Content-Type": "application/json",
        ...headers,
      },
    });
  } catch {
    throw new SkillLearningApiError(
      "Skill 提案 API に接続できません。",
      0,
      "network",
    );
  }

  if (!response.ok) {
    throw await apiError(response);
  }
  if (response.status === 204) {
    return undefined as T;
  }
  return response.json() as Promise<T>;
}

export function proposalRevisionToken(
  proposal: SkillProposal,
): SkillProposalRevisionToken {
  return {
    updated_at: proposal.updated_at,
    status: proposal.status,
    base_hash: proposal.base_hash,
    base_version: proposal.base_version,
    applied_hash: proposal.applied_hash,
    applied_version: proposal.applied_version,
    rolled_back_at: proposal.rolled_back_at,
  };
}

function sameRevision(
  proposal: SkillProposal,
  expected: SkillProposalRevisionToken,
): boolean {
  return (
    proposal.updated_at === expected.updated_at &&
    proposal.status === expected.status &&
    proposal.base_hash === expected.base_hash &&
    proposal.base_version === expected.base_version &&
    proposal.applied_hash === expected.applied_hash &&
    proposal.applied_version === expected.applied_version &&
    proposal.rolled_back_at === expected.rolled_back_at
  );
}

async function getProposal(
  proposalId: string,
): Promise<{ success: boolean; proposal: SkillProposal }> {
  return request(
    `/proposals/${encodeURIComponent(proposalId)}`,
  );
}

async function assertProposalRevision(
  proposalId: string,
  expected?: SkillProposalRevisionToken,
): Promise<void> {
  if (!expected) return;
  const current = (await getProposal(proposalId)).proposal;
  if (sameRevision(current, expected)) return;

  throw new SkillLearningApiError(
    "提案は別の操作で更新されています。最新状態を確認してください。",
    409,
    current.status === "stale" ? "stale" : "conflict",
    current,
  );
}

export const skillLearningApi = {
  async listProposals(
    options: {
      status?: SkillProposalStatus;
      projectId?: string;
      limit?: number;
    } = {},
  ): Promise<{ success: boolean; proposals: SkillProposal[] }> {
    const params = new URLSearchParams();
    if (options.status) params.set("status", options.status);
    if (options.projectId) params.set("project_id", options.projectId);
    params.set("limit", String(options.limit ?? 100));
    return request(`/proposals?${params.toString()}`);
  },

  getProposal,

  async createProposal(
    input: CreateSkillProposalInput,
  ): Promise<{ success: boolean; proposal: SkillProposal }> {
    return request("/proposals", {
      method: "POST",
      body: JSON.stringify(input),
    });
  },

  async reviseProposal(
    proposalId: string,
    proposedContent: SkillProposalContent,
    provenance: Record<string, unknown> = {},
    expected?: SkillProposalRevisionToken,
  ): Promise<{ success: boolean; proposal: SkillProposal }> {
    await assertProposalRevision(proposalId, expected);
    return request(`/proposals/${encodeURIComponent(proposalId)}`, {
      method: "PUT",
      body: JSON.stringify({
        proposed_content: proposedContent,
        provenance,
      }),
    });
  },

  async applyProposal(
    proposalId: string,
    expected?: SkillProposalRevisionToken,
  ): Promise<{ success: boolean; proposal: SkillProposal }> {
    await assertProposalRevision(proposalId, expected);
    return request(`/proposals/${encodeURIComponent(proposalId)}/apply`, {
      method: "POST",
    });
  },

  async rejectProposal(
    proposalId: string,
    reason?: string,
    expected?: SkillProposalRevisionToken,
  ): Promise<{ success: boolean; proposal: SkillProposal }> {
    await assertProposalRevision(proposalId, expected);
    return request(`/proposals/${encodeURIComponent(proposalId)}/reject`, {
      method: "POST",
      body: JSON.stringify({ reason: reason?.trim() || null }),
    });
  },

  async rollbackProposal(
    proposalId: string,
    expected?: SkillProposalRevisionToken,
  ): Promise<{ success: boolean; proposal: SkillProposal }> {
    await assertProposalRevision(proposalId, expected);
    return request(`/proposals/${encodeURIComponent(proposalId)}/rollback`, {
      method: "POST",
    });
  },

  async getUsageReceipt(
    receiptId: string,
  ): Promise<{ success: boolean; receipt: SkillUsageReceipt }> {
    return request(`/usage-receipts/${encodeURIComponent(receiptId)}`);
  },
};
